### PR TITLE
Refactored app initialization

### PR DESCRIPTION
**Pull request by tsareg**
In this PR I've refactored app intialization:
- extracted separate App component which incapsulates routing stuff.
- extracted separate routerMiddleware for web and native.

This has https://github.com/jitsi/jitsi-meet-react/pull/35 inside, so should be merged only after it.

### CODE REVIEW

File: features/app/constants.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Available application screens/routes.
+ * 
+ * @enum {string}
+ */
+export const APP_SCREEN = {
+    HOME: 'home',
```
Comment (lyubomir) [inline features/app/constants.js]: Our style guide requires an entity to be named with 1 name in its multiple occurrences. If HOME refers to features/welcome and its WelcomePage, the name here should be WELCOME as well.

File: features/app/components/native/App.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import React, { Component } from 'react';
+import { Provider } from 'react-redux';
+import { Navigator } from 'react-native';
+import { init, WelcomePage } from '../../../welcome';
+import { Conference } from '../../../conference';
```
Comment (lyubomir) [inline features/app/components/native/App.js]: Our coding style requires that imports are sorted by import location in alphabetical order:

../../../conference
../../../welcome

**Comment by tsareg** [comment]
@lyubomir fixed the pr

File: features/app/routerMiddleware.web.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @param {Store} store - Redux store.
+ * @returns {Function}
+ */
+export function routerMiddleware(store) {
+    return next => action => {
+        if (action.type === APP_NAVIGATE) {
+            switch (action.screen) {
+            case APP_SCREEN.WELCOME:
+                return store.dispatch(push('/'));
```
Comment (lyubomir) [inline features/app/routerMiddleware.web.js]: We talked about it in an online meeting. I'm just putting our that talk here in its context:

routerMiddleware appears to be suspiciously similar between .native.js and .web.js so I'd like us to explore the possibility to expand the abstraction in the direction of having 1 platform-independent routerMiddleware and other smaller platform-dependent functions that it calls.

**Comment by tsareg** [comment]
Updated the PR after MiddlewareRegistry was introduced in #47

File: features/app/components/native/App.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        // We started with NavigatorIOS and then switched to Navigator in order
+        // to support Android as well. In order to reduce the number of
+        // modifications, accept the same format of route definition.
+        return React.createElement(route.component, { navigator });
+    }
+}
+
+App.propTypes = {
+    store: React.PropTypes.object,
+    config: React.PropTypes.object
```
Comment (lyubomir) [inline features/app/components/native/App.js:67]: Please comply with the coding style which requires sorting by alphabetical order!

File: features/app/components/native/App.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+     * @inheritdoc
+     * @returns {ReactElement}
+     */
+    render() {
+        return (
+            <Provider store={this.props.store}>
+                <Navigator
+                    initialRoute={{
+                        title: 'Jitsi Meet',
+                        component: WelcomePage
```
Comment (lyubomir) [inline features/app/components/native/App.js:38]: Please comply with the coding style which requires sorting by alphabetical order!

File: features/app/components/web/App.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     * @returns {void}
+     */
+    _onConferenceRouteLeave() {
+        this.props.store.dispatch(destroy());
+    }
+}
+
+App.propTypes = {
+    store: React.PropTypes.object,
+    config: React.PropTypes.object
```
Comment (lyubomir) [inline features/app/components/web/App.js:92]: Please comply with the coding style which requires sorting by alphabetical order!

File: features/app/constants.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Available application screens/routes.
+ * 
+ * @enum {string}
+ */
+export const APP_SCREEN = {
+    WELCOME: 'welcome',
+    CONFERENCE: 'conference'
```
Comment (lyubomir) [inline features/app/constants.js:8]: Please comply with the coding style which requires sorting by alphabetical order!

File: features/app/native/navigationHandlers.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @param {Store} store - Redux store.
+ * @param {Object} action - Action object.
+ * @param {Navigator} action.navigator - Navigator instance.
+ * @returns {void}
+ */
+navigationHandlers[APP_SCREEN.WELCOME] = (store, action) => {
+    action.navigator.push({
+        title: 'Jitsi Meet',
+        component: WelcomePage
```
Comment (lyubomir) [inline features/app/native/navigationHandlers.js:26]: Please comply with the coding style which requires sorting by alphabetical order!

File: features/app/native/navigationHandlers.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+ * @param {Store} store - Redux store.
+ * @param {Object} action - Action object.
+ * @param {Navigator} action.navigator - Navigator instance.
+ * @param {string} action.room - Room name.
+ * @returns {void}
+ */
+navigationHandlers[APP_SCREEN.CONFERENCE] = (store, action) => {
+    action.navigator.push({
+        title: action.room,
+        component: Conference
```
Comment (lyubomir) [inline features/app/native/navigationHandlers.js:44]: Please comply with the coding style which requires sorting by alphabetical order!

File: features/app/native/navigationHandlers.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import Config from '../../../config';
```
Comment (lyubomir) [inline features/app/native/navigationHandlers.js:1]: I just don't get it why features/app (1) implements an App component which has a property config & (2) at the same time defines a feature-private function which deals with some other (global) config...

File: features/app/components/web/App.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Init JitsiMeetJS and new conference when we enter the "conference" route.
+     *
+     * @param {Object} route - Current route.
+     * @private
+     * @returns {void}
+     */
+    _onConferenceRouteEnter(route) {
+        const room = route.location.pathname.substr(1).toLowerCase();
+        this.props.store.dispatch(init(this.props.config, room));
```
Comment (lyubomir) [inline features/app/components/web/App.js:75]: I don't get it why we have the init action dispatched in the App component on Web & in navigationHandlers on mobile. To me, that's a signal of a very flimsy abstraction which doesn't abstract anything but rather moves source code from file to another.

**Pull request merged**